File: 101. Symmetric Tree.scala
/*
Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3
But the following [1,2,2,null,3,null,3] is not:
    1
   / \
  2   2
   \   \
   3    3
*/

    2
   / \
  1   3


class TreeNode(var _value: Int) {
   var value: Int = _value
   var left: TreeNode = null
   var right: TreeNode = null
   override def toString = value+ ":"+ "{"+  left +","+right+"}"
}

def isSymmetricImp(left: TreeNode,right:TreeNode): Boolean = {
    (left,right) match {
        case (null,null) => return true;
        case (null,_) => return false;
        case (_,null) => return false
        case (_,_)    => {
            if(left.value==right.value && isSymmetricImp(left.left,right.right)
                && isSymmetricImp(left.right,right.left))
                return true;
            else return false;
        } 
    }
}

def isSymmetric(root: TreeNode): Boolean = {
    if(root==null ) return true;
    return isSymmetricImp(root.left,root.right);
}

File: 101.Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None :
            return True
        return self.isSymmetric_imp(root.left,root.right)
    
    def isSymmetric_imp(self, left,right):
        if left is None and right is None:
            return True
        if left is None or right is None:
            return False
        if left.val != right.val:
            return False
        return self.isSymmetric_imp(left.left,right.right) and self.isSymmetric_imp(left.right,right.left)
